### DISCUSSION TITLE
AltCover with Cake 3.0.0

### DISCUSSION THREAD

**Discussion by tarnegur** [SteveGilham/altcover#192] [Q&A]
Hi.
I am trying to set up AltCover with xUnit and Cake 3.0.0. According to docs I thought it should work, but I get this errors:

The assembly 'AltCover.Cake, Version=8.6.0.0, Culture=neutral, PublicKeyToken=null' is referencing an older version of Cake.Core (2.0.0). 
For best compatibility it should target Cake.Core version 3.0.0.
Error: Error(s) occurred when compiling build script
error CS0103: The name 'cakeversion' does not exist in the current context
error CS0104: "Path" is an ambiguous reference between "Cake.Core.IO.Path" and "System.IO.Path".
error CS0103: The name 'reportDirectory' does not exist in the current context

In my .cake file I have this:
#addin "nuget:?package=Microsoft.TestPlatform.ObjectModel&Version=16.1.1"
#addin "nuget:?package=PowerShellStandard.Library&Version=5.1.0"
#addin "nuget:?package=altcover.api&Version=8.6.68"
#addin "nuget:?package=altcover.cake&Version=8.6.68"

public void RunTests(IEnumerable<FilePath> dlls)
{
        CleanDirectory("./TestResults");
        const string file = @"TestResults\VsTestDiag.log";
        if (FileExists(file)) DeleteFile(file);

        var today = DateTime.Today.ToString("yyyyMMdd_HHmmss");
		
		XUnit2(dlls,
			new XUnit2Settings {
                ArgumentCustomization = args => args.Append("/p:AltCover=true"),
				Parallelism = ParallelismOption.None,
				HtmlReport = true,
				XmlReport = true,
				NoAppDomain = true,
				OutputDirectory = "./TestResults",
				ReportName = $"TestResult_{today}",
		});
}

What am I doing wrong?

**Comment by SteveGilham** [comment]
If you're just manually setting the `dotnet test` command line by `args.Append("/p:AltCover=true")`, you don't need to include any AltCover packages in your script (only in the unit test project).  The API support is there to let the command line parameters be set programmatically and in a more type (and spelling) safe fashion.

As to the errors - 
The linker advisory about Cake 2.0 comes from my decision to take advantage of the flexibility of `dotnet`, and work with the oldest Cake version that also allows compatibility with the newest, so as not to unduly force my users to upgrade their build environment.
The compilation errors seem to relate to some other part of the Cake script not shown; so it is difficult to advise in specific terms.  In general though, the process goes

1) sub-class the AltCover.Cake.*Options types where you want to provide non-default values to the coverage gathering process, overriding the read-only properties that simply provide defaults, such as, in this case doing something like
```
  class MyPrepareOptions : AltCover.Cake.PrepareOptions
  {
    public override string Report => "./TestResults/coverage.xml"
  }
```
or adding extra filter definitions to exclude unwanted parts of the code under test, or overriding `CollectOptions` to set a coverage threshold, or ...

2) create an overall `AltCover.Cake.CoverageSettings` grouping together all these parameters
```
     var altcoverSettings = new AltCover.Cake.CoverageSettings {
        PreparationPhase = new MyPrepareOptions(),
        CollectionPhase = new MyCollectOptions(),
        Options = new MyTestOptions()
    };
```
where overrides are provided as desired

3) then include the settings into the test command line by - in this context where the code is by context a method of a `DotNetTestSettings` subtype - calling the `Concatenate` method on the local `ArgumentCustomization` value set up as it would have been without coverage collection
```
      ArgumentCustomization = altcoverSettings.Concatenate(ArgumentCustomization);
```
4) run `DotNetTest` on the appropriate test project with these settings.
